File: src/codegen/batch/builtins/file.py
#-------------------------------------------------
# CONSTANTS
#-------------------------------------------------

CODE = (
'''
set {0}={0}

set {0}.delete={0}.delete
set {0}.delete.__c=0
set {0}.delete.__f={0}.delete
goto {0}.delete_
:{0}.delete
if exist "%~4" (
    del "%~4
    set r=1
) else (
    set r=0
)
set %~1=%r%
goto :eof
:{0}.delete_

set {0}.exists={0}.exists
set {0}.exists.__c=0
set {0}.exists.__f={0}.exists
goto {0}.exists_
:{0}.exists
if exist "%~4" (set r=1) else (set r=0)
set %~1=%r%
goto :eof
:{0}.exists_

set {0}.read={0}.read
set {0}.read.__c=0
set {0}.read.__f={0}.read
goto {0}.read_
:{0}.read
set lf=^


set "r="
for /f "delims=" %%s in (%~4) do (set "r=!r!%%s!lf!")
set %1=%r%
goto :eof
:{0}.read_
''')

#-------------------------------------------------
# GLOBALS
#-------------------------------------------------

var = None

#-------------------------------------------------
# FUNCTIONS
#-------------------------------------------------

def emit_code(bat):
    global var
    if not var:
        var = bat.tempvar('string')
        var.name = '__file'
        bat.emit(CODE.format(var.name), 'decl')
        bat.decl_var('file', 'variable', global_scope=True).name = var.name

File: lib/file.js
/*----------------------------------------------------------
 * FUNCTIONS
 *--------------------------------------------------------*/

/**
 * Provides a file object for delete, reading from and writing to files.
 *
 * @param name      The name of the file.
 * @param writeMode The file access mode. Specify 'w' for overwrite mode or 'a'
 *                  for append mode.
 */
function file(name, writeMode) {
    this = []

    /*---------------------------------------
     * FIELDS
     *-------------------------------------*/

    this._writeMode = writeMode

    this.name = name

    /*---------------------------------------
     * FUNCTIONS
     *-------------------------------------*/

    append = function (_1, _2) {
        /* Windows */
        raw('echo %~5>>"%~4"', 'bat')
    }

    delete = function (_1) {
        /* Windows */
        raw('if exist "%~4" (', 'bat')
        raw('del "%~4"'       , 'bat')
        raw('set %~1=1'       , 'bat')
        raw(') else ('        , 'bat')
        raw('set %~1=0'       , 'bat')
        raw(')'               , 'bat')
        raw('goto :eof'       , 'bat')
    }

    exists = function (_1) {
        /* Windows */
        raw('if exist "%~4" (set %~1=1) else (set %~1=0)', 'bat')
        raw('goto :eof'                                  , 'bat')
    }

    read = function (_1) {
        /* Windows */
        raw('set lf=^'                                                        , 'bat')
        raw(''                                                                , 'bat')
        raw(''                                                                , 'bat')
        raw('set "r="'                                                        , 'bat')
        raw('for /f "usebackq delims=" %%s in ("%~4") do (set "r=!r!%%s!lf!")', 'bat')
        raw('set %1=%r%'                                                      , 'bat')
        raw('goto :eof'                                                       , 'bat')
    }

    write = function (_1, _2) {
        /* Windows */
        raw('echo %~5>"%~4"', 'bat')
    }

    /*---------------------------------------
     * METHODS
     *-------------------------------------*/

    /**
     * Deletes the file if it exists.
     *
     * @returns True if the file existed and was deleted.
     */
    this.delete = function () {
        return delete(this.name)
    }

    /**
     * Checks whether the file exists.
     *
     * @returns True if the file exists.
     */
    this.exists = function () {
        return exists(this.name)
    }

    /**
     * Reads the contents of the file.
     *
     * @returns The contents of the file.
     */
    this.read = function () {
        return read(this.name)
    }

    /**
     * Writes text to the file.
     *
     * @param text The text to write to the file.
     */
    this.write = function (text) {
             if (this._writeMode == 'a') append(this.name, text)
        else if (this._writeMode == 'w') write (this.name, text)
    }

    return this
}
